### ISSUE TITLE
Constant Value Checker: eliminate @StaticallyExecutable annotation and statically-executable.astub file

### ISSUE THREAD

**Issue by mernst** [typetools/checker-framework#1008]
The annotation `@StaticallyExecutable` is not needed.

Instead, the Constant Value Checker should use any `@Pure` annotation whose argument and return types are all:
 * primitives
 * primitive wrappers
 * Strings, CharSequence, and other immutable types to be indentified later
 * arrays of the above

Once this is done, the file `statically-executable.astub` is no longer needed.

**Comment by panacekcz** [comment]
I would like to point out differences between [`@Pure`](https://checkerframework.org/api/org/checkerframework/dataflow/qual/Pure.html) and [`@StaticallyExecutable`](https://checkerframework.org/api/org/checkerframework/common/value/qual/StaticallyExecutable.html), which I think make this change problematic or undesirable, unless the definitions are adjusted somehow.
1. [`@Deterministic`](https://checkerframework.org/api/org/checkerframework/dataflow/qual/Deterministic.html) (implied by `@Pure`) requires that the returned value is always the same according to `==`, so it shouldn't be used for most methods returning strings or arrays, which return newly allocated objects. However, for static execution, only the value of an array/string is important. So some methods that are currently `@StaticallyExecutable` should not be marked `@Pure`. 
Note however, that this observation contradicts a sentence from the [manual](https://checkerframework.org/manual/#constant-value-staticallyexecutable-annotation):
> A `@StaticallyExecutable` method must be `@Pure` (side-effect-free and deterministic).
2. `@Deterministic` allows the returned value to depend not only on method arguments, but also on static variables. However, for static execution, the returned value should be determined just by the arguments. For example, a getter for a static variable (with no parameters) could be `@Pure`, but its return value can vary at runtime so it should not be considered `@StaticallyExecutable`.

**Comment by mernst** [comment]
Thanks for the comments.  You are right that `@Pure` and `@StaticallyExecutable` are different.

> for static execution, only the value of an array/string is important.

Why is this true?  I agree it is true most of the time, but is it impossible that the value does not matter?

> some methods that are currently `@StaticallyExecutable` should not be marked @Pure

Another possibility would be that those methods should never have be marked `@StaticallyExecutable`

> a getter for a static variable (with no parameters) could be @Pure, but its
> return value can vary at runtime so it should not be considered
> `@StaticallyExecutable.`

I agree with this point.  Thanks!
This is discouraging, because it means that we still need lots of `@StaticallyExecutable` annotations, and I was hoping to avoid that.
Maybe we need a slightly different annotation to express that?


In the future, I would like to create new annotations `@DeterministicEquals` and `@PureEquals` that means the objects might be different but they are the same according to `equals()`.  This will express facts that cannot be expressed now.  Maybe it will be used even more than the current versions.
This is really a separate feature request, however.


Thanks again!

**Comment by panacekcz** [comment]
> Why is this true? I agree it is true most of the time, but is it impossible that the value does not matter?

I am sorry, but I do not understand the second question. Could you say it with different words or more detail?

What I meant to say with that sentence is that the Value checker uses the values returned from `@StaticallyExecutable` methods (to infer `@StringVal` and other annotations), but reference (in)equality of those has no effect on that. So `@Deterministic` is unnecesarily strong from the point of view of the Value checker which cannot make use of this guarantee. Is that what you are questioning or something else?

**Comment by mernst** [comment]
Thanks for the clarification.  I agree that *for the purpose of `@StringVal`*, determinism up to `==` does not matter, and determinism up to `equals()` is adequate.

For other scenarios, maybe determinism up to `equals()` is adequate or maybe it is not.
